src/libethash/internal.c

```diff
@@ -176,40 +144,50 @@ void ethash_calculate_dag_item(
 
 bool ethash_compute_full_data(
 	void* mem,
-	ethash_params const* params,
-	ethash_cache const* cache
+	uint64_t full_size,
+	ethash_light_t const light,
+	ethash_callback_t callback
 )
 {
-	if (params->full_size % (sizeof(uint32_t) * MIX_WORDS) != 0 ||
-		(params->full_size % sizeof(node)) != 0) {
+	if (full_size % (sizeof(uint32_t) * MIX_WORDS) != 0 ||
+		(full_size % sizeof(node)) != 0) {
 		return false;
 	}
+	uint64_t const max_n = full_size / sizeof(node);
 	node* full_nodes = mem;
+	double const progress_change = 1.0f / max_n;
+	double progress = 0.0f;
 	// now compute full nodes
-	for (unsigned n = 0; n != (params->full_size / sizeof(node)); ++n) {
-		ethash_calculate_dag_item(&(full_nodes[n]), n, params, cache);
+	for (uint64_t n = 0; n != max_n; ++n) {
+		if (callback &&
+			n % (max_n / 100) == 0 &&
```

Comment:
Let's divide by 16384 here, even that gives pretty frequent calls to the callback on a modern machine.